sitemap-index/sitemap-index-generator.py

```diff
@@ -0,0 +1,162 @@
+            new_branch = Branch(branch["gitBranchName"],
+                                branch.get("active", False),
+                                branch.get("publishOriginalBranchName", False),
+                                branch.get("urlSlug", None),
+                                branch.get("buildsWithSnooty", True))
+            branch_list.append(new_branch)
+        return branch_list
+
+    def export(self) -> Repo:
+        repo = Repo(
+            repoName=self.repoName,
+            branches=self.branches,
+            prefix=self.prefix,
+            baseUrl=self.baseUrl
+        )
+        return repo
+
+
+class ConstructSitemapEntry:
+    def __init__(self, data: Branch) -> None:
+        self.data = data
+
+        self.gitBranchName: str = data.gitBranchName
+        self.urlSuffix = self.derive_url_suffix()
+        self.extension = self.derive_extension()
+
+    def derive_extension(self) -> str:
+        if self.data.buildsWithSnooty:
+            return "/sitemap-0.xml"
+        return "/sitemap.xml.gz"
+
+    def derive_url_suffix(self) -> str:
+        urlSuffix: str = ""
+        if self.data.urlSlug:
+            urlSuffix = self.data.urlSlug
+            return urlSuffix
+        if self.data.publishOriginalBranchName:
+            urlSuffix = self.gitBranchName
+            return urlSuffix
+        return urlSuffix
+    
+    def export(self) -> SitemapUrlSuffix:
+        suffix = SitemapUrlSuffix(
+                gitBranchName=self.gitBranchName,
+                urlSuffix=self.urlSuffix,
+                extension=self.extension
+            )
+        return suffix
+                             
+def run_validation(data) -> tuple[bool, str]:
+    valid = True
+    if not check_type(str, data["repoName"]):
+        valid = False
+        return valid, "No repo name?!"
+    if not data.get("branches"):
+        valid = False
+        return valid, "No branch entry"
+    if not (data.get("prefix") and data["prefix"].get("dotcomprd")):
+        valid = False
+        return valid, "No dotcomprd prefix entry"
+    return valid, ""
+
+repos_branches = pymongo.MongoClient(os.environ.get('SNOOTY_CONN_STRING'))["pool"].repos_branches
+
+repos_branches_data = repos_branches.find()
+sitemap_urls: list[str] = []
+
+for r in repos_branches_data:
+    print(r["repoName"])
+    validity, message = run_validation(r)
+    if not validity:
+        print(message)
+        continue
+    # Skip repos that do not need sitemaps or whose sitemaps are horribly broken because built by legacy tooling
+    if r["repoName"] in excluded_repos:
+        print("Skipping")
+        continue
+    repo = ConstructRepo(r).export()
+
+    if repo.branches:
+        for b in repo.branches:
+            if b.active:
+                print(b.gitBranchName)
+                sitemap_suffix: SitemapUrlSuffix = ConstructSitemapEntry(b).export()
+                sitemap_url: str = repo.baseUrl + sitemap_suffix.urlSuffix + sitemap_suffix.extension
```

Comment:
this way of concatenating path components is easy to get wrong. I'd actually use `posixpath.join()` here and on line 52, e.g.

```
sitemap_url = posixpath.join(sitemap_suffix.urlSuffix, sitemap_suffix.extension)
```

and not include the slashes in those variables